daft/table/table_io.py

```diff
@@ -340,96 +345,118 @@ def read_csv(
+    part_keys_postfix_per_table: list[str | None]
+    partition_values = None
+    if partition_cols and len(partition_cols) > 0:
+
+        default_part = Series.from_pylist([partition_null_fallback])
+        split_tables, partition_values = table.partition_by_value(partition_keys=partition_cols)
+        assert len(split_tables) == len(partition_values)
+        pkey_names = partition_values.column_names()
+
+        values_string_values = []
+
+        for c in pkey_names:
+            column = partition_values.get_column(c)
+            string_names = column._to_str_values()
+            null_filled = column.is_null().if_else(default_part, string_names)
+            values_string_values.append(null_filled.to_pylist())
+
+        part_keys_postfix_per_table = []
+        for i in range(len(partition_values)):
+            postfix = "/".join(f"{pkey}={values[i]}" for pkey, values in zip(pkey_names, values_string_values))
+            part_keys_postfix_per_table.append(postfix)
+        tables_to_write = split_tables
+    else:
+        tables_to_write = [table]
+        part_keys_postfix_per_table = [None]
+
+    visited_paths = []
+    partition_idx = []
+
+    execution_config = get_context().daft_execution_config
+
+    TARGET_ROW_GROUP_SIZE = execution_config.parquet_target_row_group_size
+
+    if file_format == FileFormat.Parquet:
         format = pads.ParquetFileFormat()
+        inflation_factor = execution_config.parquet_inflation_factor
+        target_file_size = execution_config.parquet_target_filesize
         opts = format.make_write_options(compression=compression)
-    elif file_format == "csv":
+    elif file_format == FileFormat.Csv:
         format = pads.CsvFileFormat()
         opts = None
         assert compression is None
+        inflation_factor = execution_config.csv_inflation_factor
+        target_file_size = execution_config.csv_target_filesize
     else:
         raise ValueError(f"Unsupported file format {file_format}")
 
-    visited_paths = []
+    for i, (tab, pf) in enumerate(zip(tables_to_write, part_keys_postfix_per_table)):
+        full_path = resolved_path
+        if pf is not None and len(pf) > 0:
+            full_path = f"{full_path}/{pf}"
+
+        # TODO: For overwriting behavior, check here if dir exists to determine to delete files
+        # fs.create_dir(full_path)
+
+        arrow_table = tab.to_arrow()
+
+        size_bytes = arrow_table.nbytes
+
+        target_num_files = max(math.ceil(size_bytes / target_file_size / inflation_factor), 1)
+        num_rows = len(arrow_table)
+
+        rows_per_file = max(math.ceil(num_rows / target_num_files), 1)
+
+        target_row_groups = max(math.ceil(size_bytes / TARGET_ROW_GROUP_SIZE / inflation_factor), 1)
+        rows_per_row_group = max(min(math.ceil(num_rows / target_row_groups), rows_per_file), 1)
+
+        def file_visitor(written_file):
+            visited_paths.append(written_file.path)
+            partition_idx.append(i)
```

Comment:
If `file_visitor` is not executed eagerly at the end of this loop, then this dynamic function definition will suffer from late-binding, where `i` will be bound to the value on the last execution of this loop for all partitions. Shouldn't be an issue for `pads.write_dataset()`, which I believe is eager, but might want to bind the current `i` value as a default arg in case we make this lazy in the future (e.g. in order to parallelize the writes).

```suggestion
        def file_visitor(written_file, i=i):
            visited_paths.append(written_file.path)
            partition_idx.append(i)
```

Reply:
done!